File: kafka-connect-jdbc-sink/src/main/scala/com/datamountaineer/streamreactor/connect/jdbc/sink/writer/JdbcDbWriter.scala
package com.datamountaineer.streamreactor.connect.jdbc.sink.writer

import java.sql.SQLException

import com.datamountaineer.streamreactor.connect.jdbc.sink.StructFieldsDataExtractor
import com.datamountaineer.streamreactor.connect.jdbc.sink.config.{ErrorPolicyEnum, JdbcSinkSettings}
import com.datamountaineer.streamreactor.connect.sink._
import com.typesafe.scalalogging.slf4j.StrictLogging
import com.zaxxer.hikari.{HikariConfig, HikariDataSource}
import org.apache.kafka.connect.sink.SinkRecord

import scala.util.Try

/**
  *
  * Responsible for taking a sequence of SinkRecord and write them to the database
  *
  * @param connectionStr       - The database connection string
  * @param statementBuilder    - Returns a sequence of PreparedStatement to process
  * @param errorHandlingPolicy - An instance of the error handling approach
  */
case class JdbcDbWriter(connectionStr: String,
                        statementBuilder: PreparedStatementBuilder,
                        errorHandlingPolicy: ErrorHandlingPolicy) extends DbWriter with StrictLogging {

  private val config = new HikariConfig()
  config.setJdbcUrl(connectionStr)
  //config.setUsername("bart");
  //config.setPassword("51mp50n");
  private val ds = new HikariDataSource(config)
  //val connection = DriverManager.getConnection(connectionStr)

  /**
    * Writes the given records to the database
    *
    * @param records - The sequence of records to insert
    */
  override def write(records: Seq[SinkRecord]): Unit = {
    if (records.isEmpty)
      logger.warn("Received empty sequence of SinkRecord")
    else {

      val connection = ds.getConnection
      try {
        val statemens = statementBuilder.build(records)(connection)
        if (statemens.nonEmpty) {
          try {
            //begin transaction
            connection.setAutoCommit(false)
            statemens.foreach { statement =>
              statement.execute()
            }
            //commit the transaction
            connection.commit()
          }
          catch {
            case sqlException: SQLException =>
              logger.error(
                s"""
                   | Following error has occurred inserting data starting at
                   | topic:${records.head.topic()}
                   | offset:${records.head.kafkaOffset()}
                   | partition:${records.head.kafkaPartition()}""".stripMargin)

              //rollback the transaction
              connection.rollback()
              errorHandlingPolicy.handle(records, sqlException)(connection)
          }
          finally {
            Try {
              statemens.foreach(_.close())
            }
          }
        }
      }
      finally {
        if (connection == null) {
          connection.close()
        }
      }
    }
  }


  override def close(): Unit = {
    ds.close()
  }
}

object JdbcDbWriter {
  def apply(settings: JdbcSinkSettings): JdbcDbWriter = {
    val fieldsValuesExtractor = StructFieldsDataExtractor(settings.fields.includeAllFields, settings.fields.fieldsMappings)
    val statementBuilder = if (settings.batching) {
      BatchedPreparedStatementBuilder(settings.tableName, fieldsValuesExtractor)
    } else {
      SinglePreparedStatementBuilder(settings.tableName, fieldsValuesExtractor)
    }

    val errorHandlingPolicy = settings.errorPolicy match {
      case ErrorPolicyEnum.NOOP => NoopErrorHandlingPolicy
      case ErrorPolicyEnum.THROW => ThrowErrorHandlingPolicy
    }

    JdbcDbWriter(
      settings.connection,
      statementBuilder,
      errorHandlingPolicy
    )
  }
}

File: kafka-connect-jdbc-sink/src/main/java/com/datamountaineer/streamreactor/connect/jdbc/sink/writer/JdbcDbWriter.java
/**
 * Copyright 2015 Datamountaineer.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package com.datamountaineer.streamreactor.connect.jdbc.sink.writer;

import com.datamountaineer.streamreactor.connect.jdbc.sink.config.JdbcSinkSettings;
import com.datamountaineer.streamreactor.connect.sink.DbWriter;
import com.google.common.collect.Iterators;
import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import org.apache.kafka.connect.sink.SinkRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Collection;
import java.util.List;

/**
 * Responsible for taking a sequence of SinkRecord and write them to the database
 */
public final class JdbcDbWriter implements DbWriter {
    private static final Logger logger = LoggerFactory.getLogger(JdbcDbWriter.class);

    private final PreparedStatementBuilder statementBuilder;
    private final ErrorHandlingPolicy errorHandlingPolicy;

    //provides connection pooling
    private final HikariDataSource dataSource;

    /**
     * @param connectionStr       - The database connection string
     * @param statementBuilder    - Returns a sequence of PreparedStatement to process
     * @param errorHandlingPolicy - An instance of the error handling approach
     */
    public JdbcDbWriter(String connectionStr, PreparedStatementBuilder statementBuilder, ErrorHandlingPolicy errorHandlingPolicy) {

        final HikariConfig config = new HikariConfig();
        config.setJdbcUrl(connectionStr);
        this.dataSource = new HikariDataSource(config);
        this.statementBuilder = statementBuilder;
        this.errorHandlingPolicy = errorHandlingPolicy;
    }

    /**
     * Writes the given records to the database
     *
     * @param records - The sequence of records to insert
     */
    @Override
    public void write(final Collection<SinkRecord> records) {
        if (records.isEmpty()) {
            logger.warn("Received empty sequence of SinkRecord");
        }
        else {

            Connection connection = null;
            List<PreparedStatement> statements = null;
            try {
                connection = dataSource.getConnection();
                statements = statementBuilder.build(records, connection);
                if (!statements.isEmpty()) {
                    //begin transaction
                    connection.setAutoCommit(false);
                    for (final PreparedStatement statement : statements) {
                        if (statementBuilder.isBatching()) {
                            statement.executeBatch();
                        } else {
                            statement.execute();
                        }
                    }
                    //commit the transaction
                    connection.commit();
                }
            } catch (SQLException sqlException) {
                final SinkRecord firstRecord = Iterators.getNext(records.iterator(), null);
                logger.error(String.format("Following error has occurred inserting data starting at topic:%s offset:%d partition:%d",
                        firstRecord.topic(),
                        firstRecord.kafkaOffset(),
                        firstRecord.kafkaPartition()));

                if (connection != null) {//rollback the transaction
                    try {
                        connection.rollback();
                    } catch (Throwable t) {
                    }
                }

                //handle the exception
                errorHandlingPolicy.handle(records, sqlException, connection);
            } finally {
                if (statements != null) {
                    for (final PreparedStatement statement : statements) {
                        try {
                            statement.close();
                        } catch (Throwable t) {
                        }
                    }
                }

                if (connection != null) {
                    try {
                        connection.close();
                    } catch (Throwable t) {
                    }
                }
            }
        }
    }

    @Override
    public void close() {
        dataSource.close();
    }

    /**
     * Creates an instance of JdbcDbWriter from the Jdbc sink settings.
     *
     * @param settings - Holds the sink settings
     * @return Returns a new instsance of JdbcDbWriter
     */
    public static JdbcDbWriter from(final JdbcSinkSettings settings) {
        final PreparedStatementBuilder statementBuilder = PreparedStatementBuilderHelper.from(settings);
        final ErrorHandlingPolicy errorHandlingPolicy = ErrorHandlingPolicyHelper.from(settings.getErrorPolicy());
        return new JdbcDbWriter(settings.getConnection(), statementBuilder, errorHandlingPolicy);
    }

    /**
     * Get the prepared statement builder
     * @return a PreparedStatementBuilder
     * */
    public PreparedStatementBuilder getStatementBuilder() {
        return statementBuilder;
    }

    /**
     * Get the Error Handling Policy for the task
     *
     * @return A ErrorHandlingPolicy
     * */
    public ErrorHandlingPolicy getErrorHandlingPolicy() {
        return errorHandlingPolicy;
    }
}